---
title: "R Notebook"
output: html_notebook
---

```{r}
library(dplyr)
library(dplyr)
library(ggplot2)
library(reshape2)
library(tidyr)
library(caret)


OTUtoMAG <- read.csv("../data/VsearchMap6Out99.tsv", sep = "\t", header = F, check.names = F) %>% select(V1, V2)
OTUtoMAG$V2 = gsub("::.*", "", OTUtoMAG$V2)
rxn_annot = readRDS("../data/df_rxnAnnotation.rds")


two_rxn <- read.csv("../data/rxn_correlations_2M.csv")
two_sig_rxn <- two_rxn %>% filter(p.adj < 0.05)
two_sig_rxn  = merge(two_sig_rxn, rxn_annot, by.x = "OTU", by.y = 0) %>% select(-RxnID)


nine_rxn <- read.csv("../data/rxn_correlations_9M.csv")
nine_sig_rxn <- nine_rxn %>% filter(p.adj < 0.05)
fifteen_rxn <- read.csv("../data/rxn_correlations_15M.csv")
fifteen_sig_rxn <- fifteen_rxn %>% filter(p.adj < 0.05)
twentyfour_rxn <- read.csv("../data/rxn_correlations_24M.csv")
twentyfour_sig_rxn <- twentyfour_rxn %>% filter(p.adj < 0.05)
thirty_rxn <- read.csv("../data/rxn_correlations_30M.csv")
thirty_sig_rxn <- thirty_rxn %>% filter(p.adj < 0.05)


### load normalized reaction data
cecumRxNCounts = as.matrix(read.csv("../data/cecum_rxn_abundance.csv", check.names = F, row.names = 1))
cecumRxNCounts <- apply(cecumRxNCounts, 2, function(col) {col/sum(col)})
cecumRxNCounts = cecumRxNCounts[-nearZeroVar(t(cecumRxNCounts)), ]


colonRxNCounts = as.matrix(read.csv("../data/colon_rxn_abundance.csv", check.names = F, row.names = 1))
colonRxNCounts <- apply(colonRxNCounts, 2, function(col) {col/sum(col)})
colonRxNCounts = colonRxNCounts[-nearZeroVar(t(colonRxNCounts)), ]

stoolRxNCounts = as.matrix(read.csv("../data/stool_rxn_abundance.csv", check.names = F, row.names = 1))
stoolRxNCounts <- apply(stoolRxNCounts, 2, function(col) {col/sum(col)})
stoolRxNCounts = stoolRxNCounts[-nearZeroVar(t(stoolRxNCounts)), ]

metadata <- read.csv("../data/Jena_mouse_clean_RNA.csv")


getTableForOtu <- function(counts1, counts2, counts3, otu, age = 2)
{
    current_otu = as.data.frame(counts1[otu, ])
    current_otu2 = merge(current_otu,  as.data.frame(counts2[otu, ]) , by.x = 0, by.y = 0 )
    current_otu3 = merge(current_otu2,  as.data.frame(counts3[otu, ]) , by.x = "Row.names", by.y = 0 )
    colnames(current_otu3) <- c("patID", "cecum_count", "colon_count", "stool_count")
    rownames(current_otu3) <- gsub(".*/", "", current_otu3$patID)
    temp_meta = metadata %>% select(c(Age, ID))
    temp_meta$ID =  gsub("_.", "", temp_meta$ID)
    temp_meta = unique(temp_meta)
    current_otu4 = merge(current_otu3, temp_meta,  by.x = 0, by.y = "ID")
    current_otu4 = current_otu4[current_otu4$Age == age, ]
    rownames(current_otu4) = current_otu4$Row.names
    current_otu4 = current_otu4[-c(1)]
    colnames(current_otu4) <- c("ID", "1_","2_","3_","Age")
    long_data <- pivot_longer(
      data = current_otu4,
      cols = ends_with("_"),
      names_to = "Organ",
      values_to = "Count"
    )
    return(long_data)
}

t <- getTableForOtu(cecumRxNCounts, colonRxNCounts, stoolRxNCounts, "rxn00473_c0")
trp1= ggplot(t, aes(x = Organ, y = Count)) +
  geom_boxplot() +
  labs(title = "Tryptophanase", x = "Groups", y = "Values")

enrichment_map = readRDS("../data/subsysMap-cognition-cohort.RDS")
all_cecum_reaction_names = rownames(cecumRxNCounts)
all_colon_reaction_names = rownames(colonRxNCounts)
all_stool_reaction_names = rownames(stoolRxNCounts)

# I think because we are drawing from all three sites we have to take the union as a universe
all_reaction_names = union(all_cecum_reaction_names, all_colon_reaction_names)
all_reaction_names = union(all_reaction_names, all_stool_reaction_names)

## Alternatively use FVA reactions
obj_metabolicModelsMouse <- readRDS("../data/fva_99_reactions_thr06_MetamouseHQBins20211001.RDS")


##Get list of all reactions contained in the models
lst_allRxns <- c()
for(obj_model in obj_metabolicModelsMouse){
  lst_allRxns <- union(lst_allRxns, obj_model$active)
}

# All reactions taht have a subsystem reaction, note that these are not all of them.
# We lost 2173 - 2124 
universe = lst_allRxns#Amount of all active reactions  #enrichment_map[names(enrichment_map) %in% all_reaction_names]
# Now the we assign subsystems to the significant reaction 
up_2M = two_sig_rxn[two_sig_rxn$coefficient > 0, ]
down_2M = two_sig_rxn[two_sig_rxn$coefficient < 0, ]


up_9M = nine_sig_rxn[nine_sig_rxn$coefficient > 0, ]
down_9M = nine_sig_rxn[nine_sig_rxn$coefficient < 0, ]

up_15M = fifteen_sig_rxn[fifteen_sig_rxn$coefficient > 0, ]
down_15M = fifteen_sig_rxn[fifteen_sig_rxn$coefficient < 0, ]
# calculate significantly enriched subsytems
calc_terms <- function(up_down) {
significant_reactions = enrichment_map[names(enrichment_map) %in% up_down$OTU]
all_terms = unique(unlist(significant_reactions))
# Count is sig. in-term
pvals = data.frame(name = as.character(), pval = as.numeric(), count = as.integer())
# We want to test all terms for enrichment #TODO Figure out if those should only be the sign.
for(t in all_terms) {
  #t = "pectin degradation II"
  upper_left  = 0
  lower_left  = 0
  upper_right  = 0
  lower_right  = 0
  # Go through all sig. reaction and count how of t occurs as a term
  for(r in names(significant_reactions))
  {
    # Get the associated terms
    sig_terms = enrichment_map[[r]]
    if(t %in% sig_terms) {
      upper_left = upper_left +  1
    }
  }
  
  non_sig = setdiff(universe, names(significant_reactions))
  # Go through all non sig. reactions and count how of t does not occurs as a term
  # CHECK IF lower_left+upper_left equals the occurences of the term in stefanos model
  # NOT IN THE translation map 
  for(r in  non_sig)
  {
    terms = enrichment_map[[r]]
    if((t %in% terms)) {
      upper_right = upper_right +  1
    }
  }
  
  # Go through all sig. reactions and count how of t does not occur as a term
  non_sig = setdiff(universe, names(significant_reactions))
  for(r in names(significant_reactions))
  {
    terms = enrichment_map[[r]]
    if(!(t %in% terms)) {
      lower_left = lower_left +  1
    }
  }
  # Go through all non-sig. reactions and count how of t does not occur as a term
  for(r in non_sig )
  {
    terms = enrichment_map[[r]]
    if(!(t %in% terms)) {
      lower_right = lower_right +  1
    }
  }
 # print(t)
  
  totalA1 = upper_left + upper_right
  totalA = upper_left + lower_left
  totalB = upper_right + lower_right
  #cat(upper_left, " ", upper_right, "\n")
  #cat(lower_left,  " ", lower_right, "\n")
  #cat("----------------------------------", "\n")
  cont = matrix(c(upper_left, lower_left, upper_right, lower_right), nrow = 2, ncol = 2)
  #cat("+++++++++++++++++++++++++++++++++++++++", "\n")
  # TODO ONE SIDED OR TWO SIDE. PRETTY SURE IT SHOULD BE ONE SIDED
  pp_value2 = fisher.test(cont, one)
  #phyper(upper_left-1, totalA1 , totalA1 - totalA, totalB, lower.tail = FALSE )
  pvals[nrow(pvals) + 1, ] = c(t, pp_value2$p.value, upper_left)
}
pvals$p.adj = p.adjust(pvals$pval, method = "BH")
pvals = pvals[pvals$p.adj < 0.05, ]
return(pvals)
}
terms_up_2M = calc_terms(up_2M)
terms_down_2M = calc_terms(down_2M)
write.csv(terms_up_2M, "../data/upregulated_terms_2M.csv")
write.csv(terms_down_2M, "../data/downregulated_terms_2M.csv")
both = intersect(terms_down_2M$name, terms_up_2M$name)
write.csv(both, "../data/up_and_down_terms_2M.csv")


terms_up_9M = calc_terms(up_9M)
terms_down_9M = calc_terms(down_9M)
write.csv(terms_up_9M, "../data/upregulated_terms_9M.csv")
write.csv(terms_down_9M, "../data/downregulated_terms_9M.csv")
both = intersect(terms_down_9M$name, terms_up_9M$name)
write.csv(both, "../data/up_and_down_terms_9M.csv")

terms_up_15M = calc_terms(up_15M)
terms_down_15M = calc_terms(down_15M)
write.csv(terms_up_15M, "../data/upregulated_terms_15M.csv")
write.csv(terms_down_15M, "../data/downregulated_terms_15M.csv")
both = intersect(terms_down_15M$name, terms_up_15M$name)
write.csv(both, "../data/up_and_down_terms_15M.csv")




```

```{r}
library(ggplot2)
ggplot(terms_up_2M, aes(x = name, y = count)) +
  geom_bar(stat = "identity", fill = "blue", width = 0.8, position = position_dodge2(2)) + 
  coord_flip() +
  theme_minimal() +
  labs(
    title = "Barplot Example",
    x = "Categories",
    y = "Values"
  )
ggsave("test.png", height = 14)
```


```{r}
create_network_reac_overlap = function(data)
{
  enrichment_map = readRDS("../data/subsysMap-cognition-cohort.RDS")
  map = data.frame(rxn = as.character(), subs = as.character())
  obj_metabolicModelsMouse <- readRDS("../data/fva_99_reactions_thr06_MetamouseHQBins20211001.RDS")
  ##Get list of all reactions contained in the models
  lst_allRxns <- c()
  for(obj_model in obj_metabolicModelsMouse){
    lst_allRxns <- union(lst_allRxns, obj_model$active)
  }
  # Go through all sign reaction
  for(t in data$name)
  {
    # DO I WANT TO INCLUDE ALL REACTIONS THAT ARE IN A SUBSYSTEM?, maybe just active ones
    for(r in names(enrichment_map))
    {
      # Check if the current reaction is the current subsystem, if yes
      if(t %in% enrichment_map[[r]])
      {
        # Make an entry that list r as a member of t
        if( r %in% lst_allRxns)
        {
          map[nrow(map) +1, ] = c(r, t)
        }
      }
    }
  }
  
  # Create overlap matrix
  subs =unique(map$subs)
  subs = gsub(",", ":", subs)
  overlap = matrix(nrow = length(subs), ncol = length(subs))
  rownames(overlap) = subs
  colnames(overlap) = subs
  c  = 0
  for(s1 in subs) {
    for(s2 in subs) {
      r1 = map[map$subs == s1,]
      r2 = map[map$subs == s2,]
      overlap[s1, s2] = length(intersect(r1$rxn, r2$rxn))
      if(s1 == s2)
      { 
        overlap[s1, s2]  = 0
      }
    }
  }
  return(overlap)
}

data = read.csv("../data/upregulated_terms_2M.csv")
up_network_2M = create_network_reac_overlap(data)
write.csv(up_network_2M, "../data/cytoscpae/up_network_2M.mat",row.names = F)

data = read.csv("../data/downregulated_terms_2M.csv")
down_network_2M = create_network_reac_overlap(data)
write.csv(down_network_2M, "../data/cytoscpae/down_network_2M.mat",row.names = F)

data = read.csv("../data/upregulated_terms_9M.csv")
up_network_9M = create_network_reac_overlap(data)
write.csv(up_network_9M, "../data/cytoscpae/up_network_9M.mat",row.names = F)

data = read.csv("../data/downregulated_terms_9M.csv")
down_network_9M = create_network_reac_overlap(data)
write.csv(down_network_9M, "../data/cytoscpae/down_network_9M.mat",row.names = F)
```

```{r}
library(zoom)
#overlap[overlap L> ] <- 0
# SHARED REACTIONS BETWEEN SUBSYSTEMS
g <- graph_from_adjacency_matrix(up_network_2M, mode = "undirected", weighted = TRUE)
layout <- layout.auto(g)
node.size= c(10)
plot(g, layout = layout, vertex.size=node.size, vertex.label.cex = 0.6)
```


```{r}
library(igraph)

g = graph_from_data_frame(map)

layout <- layout.norm(layout)

plot(g, layout = layout_with_kk, edge.arrow.size = 0.1, vertex.size = 0.2, vertex.label.dist = 2, vertex.label.cex = 0.7, edge.label = 0)
```


```{r}
# "aerobic respiration I (cytochrome c)"
reacs = map[map$subs == "aerobic respiration I (cytochrome c)", ]
# use only the sig. assoc reactions
reacs = reacs[reacs$rxn %in% two_sig_rxn$OTU, ]
reacs = merge(reacs, two_sig_rxn, by.x  = "rxn", by.y = "OTU")
t <- getTableForOtu(cecumRxNCounts, colonRxNCounts, stoolRxNCounts, "rxn08976_c0")
trp1= ggplot(t, aes(x = Organ, y = Count)) +
  geom_boxplot() +
  labs(title = "Tryptophanase", x = "Groups", y = "Values")
trp1
```
  
```{r}
library(patchwork)
r <- getTableForOtu(cecumRxNCounts, colonRxNCounts, stoolRxNCounts, "EX_cpd00065_e0")
trp2= ggplot(r, aes(x = Organ, y = Count)) +
  geom_boxplot() +
  labs(title = "Tryptophan excretion", x = "Groups", y = "Values")
trp1 | trp2
```


```{r}
amount_sig_reac = data.frame(count_pos = numeric(), count_neg = numeric(),  age= numeric())
amount_sig_reac[nrow(amount_sig_reac) + 1, ] = c(nrow(two_sig_rxn[two_sig_rxn$coefficient >= 0, ]),nrow(two_sig_rxn[two_sig_rxn$coefficient < 0, ]), 2)

amount_sig_reac[nrow(amount_sig_reac) + 1, ] = c(nrow(nine_sig_rxn[nine_sig_rxn$coefficient >= 0, ]), nrow(nine_sig_rxn[nine_sig_rxn$coefficient < 0, ]), 9)

#amount_sig_reac = as.numeric(amount_sig_reac$value)

amount_sig_reac[nrow(amount_sig_reac) + 1, ] = c(nrow(fifteen_sig_rxn[fifteen_sig_rxn$coefficient >= 0, ]), nrow(fifteen_sig_rxn[fifteen_sig_rxn$coefficient < 0, ]), 15)


amount_sig_reac[nrow(amount_sig_reac) + 1, ] = c(nrow(twentyfour_sig_rxn[twentyfour_sig_rxn$coefficient >= 0, ]), nrow(twentyfour_sig_rxn[twentyfour_sig_rxn$coefficient < 0, ]), 24)


amount_sig_reac[nrow(amount_sig_reac) + 1, ] = c(nrow(thirty_sig_rxn[thirty_sig_rxn$coefficient >= 0, ]), nrow(thirty_sig_rxn[thirty_sig_rxn$coefficient < 0, ]), 30)

amount_sig_reac_long <- melt(amount_sig_reac, id.vars = "age")
ggplot(amount_sig_reac_long, aes(x = age, y = value, fill = variable)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("count_pos" = "blue", "count_neg" = "orange")) +
  scale_x_discrete(
    limits = c(2, 9, 15, 24, 30) # Specify the desired order of categories
  )  + 
  theme_minimal()
```



```{r}
OTUtoMAG <- read.csv("../data/VsearchMap6Out99.tsv", sep = "\t", header = F, check.names = F) %>% select(V1, V2)
OTUtoMAG$V2 = gsub("::.*", "", OTUtoMAG$V2)
two_mag = read.csv("../data/mag_correlation_2M.csv", row.names = 1)
nine_mag = read.csv("../data/mag_correlation_9M.csv", row.names = 1)
fifteen_mag = read.csv("../data/mag_correlation_15M.csv", row.names = 1)
twentyfour_mag = read.csv("../data/mag_correlation_24M.csv", row.names = 1)
thirty_mag = read.csv("../data/mag_correlation_30M.csv", row.names = 1)

taxonomyHQ = read.csv("../data/TaxonomyHQDraftMAGs.csv", sep = "\t")
taxonomyHQ = taxonomyHQ %>% select(c(1 : 8))

two_mag_sig = two_mag %>% filter(p.adj < 0.05)
two_mag_sig = merge(two_mag_sig, OTUtoMAG, by.x = "OTU", by.y = "V1" )
two_mag_sig = merge(two_mag_sig, taxonomyHQ, by.x = "V2", by.y = "user_genome" )
nine_mag_sig = nine_mag %>% filter(p.adj < 0.05)
nine_mag_sig = merge(nine_mag_sig, OTUtoMAG, by.x = "OTU", by.y = "V1" )
nine_mag_sig = merge(nine_mag_sig, taxonomyHQ, by.x = "V2", by.y = "user_genome" )
fifteen_mag_sig = fifteen_mag %>% filter(p.adj < 0.05)
fifteen_mag_sig = merge(fifteen_mag_sig, OTUtoMAG, by.x = "OTU", by.y = "V1" )
fifteen_mag_sig = merge(fifteen_mag_sig, taxonomyHQ, by.x = "V2", by.y = "user_genome" )
twentyfour_mag_sig = twentyfour_mag %>% filter(p.adj < 0.05)
twentyfour_mag_sig = merge(twentyfour_mag_sig, OTUtoMAG, by.x = "OTU", by.y = "V1" )
twentyfour_mag_sig = merge(twentyfour_mag_sig, taxonomyHQ, by.x = "V2", by.y = "user_genome" )
thirty_mag_sig = thirty_mag %>% filter(p.adj < 0.05)
thirty_mag_sig = merge(thirty_mag_sig, OTUtoMAG, by.x = "OTU", by.y = "V1" )
thirty_mag_sig = merge(thirty_mag_sig, taxonomyHQ, by.x = "V2", by.y = "user_genome" )

write.csv(two_mag_sig, "../data/sig_mags_2M.csv")
write.csv(nine_mag_sig, "../data/sig_mags_9M.csv")
write.csv(fifteen_mag_sig, "../data/sig_mags_15M.csv")
write.csv(twentyfour_mag_sig, "../data/sig_mags_24M.csv")
write.csv(thirty_mag_sig, "../data/sig_mags_30M.csv")

```

```{r}
library(vegan)
source("utils.R")

cecum_mag_counts = t(read.table("../data/cecum_mag_counts.csv", sep = ",", header = T, row.names = 1))
cecum_16S_counts = t(filter_per_organ(metadata, count_data, "Cecum", normalize = F, map_to_mags = F))



colon_mag_counts = t(read.table("../data/colon_mag_counts.csv", sep = ",", header = T, row.names = 1))
colon_16S_counts = t(filter_per_organ(metadata, count_data, "Colon", normalize = F, map_to_mags = F))

mag_vs_16s_alpha = function(mag_counts, sixteenS_counts, organ) {
  mag_shannon <- as.data.frame(diversity(mag_counts, index = "shannon"))                          
  mag_shannon$grp = "MAG"
  ss_shannon <- as.data.frame(diversity(sixteenS_counts, index = "shannon"))                          
  ss_shannon$grp = "16S"
  colnames(mag_shannon)= c("alpha", "grp")
  colnames(ss_shannon)= c("alpha", "grp")
  overall_shannon = rbind(mag_shannon, ss_shannon)
  p<-ggplot(overall_shannon, aes(y=alpha, x= grp)) + 
    geom_boxplot() +
    ggtitle(paste("16S vs MAG abundances across all samples across all ages in ", organ)) 
  return(p)
}

cecum_16s_vs_mag = mag_vs_16s_alpha(cecum_mag_counts, cecum_16S_counts, "cecum")
cecum_16s_vs_mag

colon_16s_vs_mag = mag_vs_16s_alpha(colon_mag_counts, colon_16S_counts, "colon")
colon_16s_vs_mag


colon_16s_vs_mag = mag_vs_16s_alpha(colon_mag_counts, colon_16S_counts, "colon")
colon_16s_vs_mag
# calculate Shannon index using vegan package
```

```{r}
library(vegan)
library(ggplot2)
library(patchwork)
source("utils.R")

cecum_mag_2M_counts  = load_and_calc("../data/cecum_mag_counts.csv", "2\\.", 1)
colon_mag_2M_counts  = load_and_calc("../data/colon_mag_counts.csv", "2\\.", 2)
stool_mag_2M_counts  = load_and_calc("../data/stool_mag_counts.csv", "2\\.", 3)
overall_mag_2M_counts = rbind(cecum_mag_2M_counts, colon_mag_2M_counts,stool_mag_2M_counts)
p1<-ggplot(overall_mag_2M_counts, aes(y=alpha, group= grp)) + 
  geom_boxplot() +
  ggtitle("MAG abundances")

cecum_16S_counts = load_and_calc(as.data.frame(t(filter_per_organ(metadata, count_data, "Cecum", normalize = F, map_to_mags = F))), "2/", 1)
colon_16S_counts = load_and_calc(as.data.frame(t(filter_per_organ(metadata, count_data, "Colon", normalize = F, map_to_mags = F))), "2/", 2)
stool_16S_counts =  load_and_calc(as.data.frame(t(filter_per_organ(metadata, count_data, "stool", normalize = F, map_to_mags = F))), "2/", 3)
overall_16S_counts = rbind(cecum_16S_counts, colon_16S_counts,stool_16S_counts)
rm(cecum_16S_counts, colon_16S_counts,stool_16S_counts)
rm(cecum_mag_2M_counts, colon_mag_2M_counts,stool_mag_2M_counts)

p2<-ggplot(overall_16S_counts, aes(y=alpha, group= grp)) + 
  geom_boxplot() + 
  ggtitle("16S Abundances")
p3 = p1 | p2 +  plot_annotation(title = "2M alpha-diversity along cecum, prox. col., dist. col. ")
print(p3)

## 16S data shows a diversity trajectory that is more in line with literature
```
```{r}
# OTU OverlapM
library(VennDiagram)
source("utils.R")
otus_in_cecum = rownames(filter_per_organ(metadata, count_data, "Cecum", normalize = F, map_to_mags=F)  %>%
    mutate(all_positive = rowSums(. > 0) == ncol(.)) %>%
    filter(all_positive == F))

otus_in_colon = rownames(filter_per_organ(metadata, count_data, "Colon", normalize = F, map_to_mags = F) %>%
    mutate(all_positive = rowSums(. > 0) == ncol(.)) %>% 
    filter(all_positive == F))

otus_in_stool = rownames(filter_per_organ(metadata, count_data, "stool", normalize = F, map_to_mags = F) %>%
    mutate(all_positive = rowSums(. > 0) == ncol(.)) %>% 
    filter(all_positive == F))

sym_diff <- function(a,b) setdiff(union(a,b), intersect(a,b))
sym_diff(otus_in_cecum, otus_in_stool)
library(ggvenn)
x = list(Cecum = otus_in_cecum, Proximal = otus_in_colon, Distal = otus_in_stool)
ggvenn(
  x, 
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF"),
  stroke_size = 0.5, set_name_size = 4,
  show_percentage = F
  )
```

```{r}
cecumRxNCounts = as.matrix(read.csv("../data/cecum_rxn_abundance.csv", check.names = F, row.names = 1))
filtered_df_cecum <- as.data.frame(cecumRxNCounts) %>%
     filter_all(any_vars(. != 0))
colonRxNCounts = as.matrix(read.csv("../data/colon_rxn_abundance.csv", check.names = F, row.names = 1))
filtered_df_colon <- as.data.frame(colonRxNCounts) %>%
     filter_all(any_vars(. != 0))
stoolRxNCounts = as.matrix(read.csv("../data/stool_rxn_abundance.csv", check.names = F, row.names = 1))
filtered_df_stool <- as.data.frame(stoolRxNCounts) %>%
     filter_all(any_vars(. != 0))
present_reactions_cecum = rownames(filtered_df_cecum)
present_reactions_colon = rownames(filtered_df_colon)
present_reactions_stool = rownames(filtered_df_stool)




present_reactions_ONLY_in_proximal =as.data.frame((setdiff(present_reactions_colon, union(present_reactions_cecum, present_reactions_stool))))
present_reactions_IN_proximal_and_distal = as.data.frame(setdiff(intersect(present_reactions_colon, present_reactions_stool), present_reactions_cecum))

colnames(present_reactions_ONLY_in_proximal) = c("rxn")
colnames(present_reactions_IN_proximal_and_distal) = c("rxn")
rxn_annot = readRDS("../data/df_rxnAnnotation.rds")
present_reactions_ONLY_in_proximal = merge(present_reactions_ONLY_in_proximal, rxn_annot, by.x = "rxn", by.y = "RxnID", all.x =T)
present_reactions_IN_proximal_and_distal = merge(present_reactions_IN_proximal_and_distal, rxn_annot, by.x = "rxn", by.y = "RxnID", all.x =T)
library(ggvenn)
x = list(Cecum = present_reactions_cecum, Proximal = present_reactions_colon, Distal = present_reactions_stool)
ggvenn(
  x, 
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF"),
  stroke_size = 0.5, set_name_size = 4,
  show_percentage = F
  )
```

```{r}
source("utils.R")
cecum_16S_counts = as.data.frame(t(filter_per_organ(metadata, count_data, "Cecum", normalize = T, map_to_mags = F)))
cecum_16S_counts$group = 1
colon_16S_counts = as.data.frame(t(filter_per_organ(metadata, count_data, "Colon", normalize = T, map_to_mags = F)))
colon_16S_counts$group = 2
stool_16S_counts = as.data.frame(t(filter_per_organ(metadata, count_data, "stool", normalize = T, map_to_mags = F)))
stool_16S_counts$group = 3
rm(OTUtoMAG)
rm(metadata)
rm(count_data)
all.equal(colnames(cecum_16S_counts), colnames(colon_16S_counts))
all.equal(colnames(cecum_16S_counts), colnames(stool_16S_counts))
column_names =  colnames(cecum_16S_counts)
empty_df <- setNames(data.frame(matrix(ncol = length(column_names))), column_names)
empty_df = t(rbind(cecum_16S_counts, colon_16S_counts, stool_16S_counts))
groups = as.data.frame(empty_df["group", ])
# TODO FIX
#empty_df = empty_df[-nearZeroVar(t(empty_df)), ]


### OTU wise PCA
library(ggplot2)
pca_result <- prcomp(empty_df[c(1:(nrow(empty_df))-1), ], scale = T)
pca_df <- as.data.frame(pca_result$x[, 1:2])
#pca_df$group <- groups
ggplot(pca_df, aes(x = PC1, y = PC2)) +
  geom_point() +
  labs(title = "PCA Plot with Groups", x = "PC1", y = "PC2") +
  theme_minimal()







```
